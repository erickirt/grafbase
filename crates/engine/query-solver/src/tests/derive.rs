use crate::assert_solving_snapshots;

const SCHEMA: &str = r#"
    type Product
      @join__type(graph: EXT)
    {
      authorId: ID!
      commentIds: [ID!]!
      inventoryCountryId: ID!
      inventoryWarehouseId: ID!
      inventoriesKeys: [InventoryKeys!]!
      code: String!
      id: ID!
      author: User! @composite__derive(graph: EXT)
      comments: [Comment!]! @composite__derive(graph: EXT)
      inventory: Inventory! @composite__derive(graph: EXT)
      inventories: [Inventory!]! @composite__derive(graph: EXT)
    }

    type User
        @join__type(graph: EXT, key: "id", resolvable: false)
    {
      id: ID!
    }

    type Comment
        @join__type(graph: EXT, key: "id", resolvable: false)
    {
      id: ID!
    }

    type InventoryKeys @join__type(graph: EXT) {
        countryId: ID!
        warehouseId: ID!
    }

    type Inventory
        @join__type(graph: EXT, key: "countryId warehouseId", resolvable: false)
    {
        countryId: ID!
        warehouseId: ID!
    }

    type Query
    {
      products: [Product!]! @join__field(graph: EXT)
    }

    enum join__Graph
    {
    EXT @join__graph(name: "ext", url: "http://localhost:8080")
    }
"#;

#[test]
fn single_id() {
    assert_solving_snapshots!(
        "single_id",
        SCHEMA,
        r#"
        query {
            products {
                author {
                    id
                }
            }
        }
        "#
    );
}

#[test]
fn composite_keys() {
    assert_solving_snapshots!(
        "composite_keys",
        SCHEMA,
        r#"
        query {
            products {
                 inventory {
                    countryId
                    warehouseId
                }
            }
        }
        "#
    );
}

#[test]
fn batch_single_id() {
    assert_solving_snapshots!(
        "batch_single_id",
        SCHEMA,
        r#"
        query {
            products {
                comments {
                    id
                }
            }
        }
        "#
    );
}

#[test]
fn batch_composite_keys() {
    assert_solving_snapshots!(
        "batch_composite_keys",
        SCHEMA,
        r#"
        query {
            products {
                inventories {
                    countryId
                    warehouseId
                }
            }
        }
        "#
    );
}
