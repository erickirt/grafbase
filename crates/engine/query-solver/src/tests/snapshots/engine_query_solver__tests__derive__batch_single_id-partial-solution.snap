---
source: crates/engine/query-solver/src/tests/derive.rs
expression: "digraph {\n    0 [ label = \"root\" ]\n    1 [ label = \"Root#ext\", color=royalblue,shape=parallelogram ]\n    2 [ label = \"Query.products\" ]\n    3 [ label = \"Product.comments\" ]\n    4 [ label = \"Product.commentIds\" ]\n    5 [ label = \"Comment.id\" ]\n    0 -> 1 [ color=royalblue,fontcolor=royalblue ]\n    1 -> 2 [  ]\n    2 -> 3 [  ]\n    2 -> 4 [  ]\n    4 -> 3 [ color=darkorchid,arrowhead=halfopen,style=dashed ]\n    3 -> 5 [  ]\n    4 -> 5 [ color=darkorchid,arrowhead=halfopen,style=dashed ]\n}\n"
---
digraph {
    0 [ label = "root" ]
    1 [ label = "Root#ext" ]
    2 [ label = "Query.products" ]
    3 [ label = "Product.comments" ]
    4 [ label = "Product.commentIds" ]
    5 [ label = "Comment.id" ]
    0 -> 1 [ label = "QueryPartition" ]
    1 -> 2 [ label = "Field" ]
    2 -> 3 [ label = "Field" ]
    2 -> 4 [ label = "Field" ]
    4 -> 3 [ label = "Derive" ]
    3 -> 5 [ label = "Field" ]
    4 -> 5 [ label = "Derive" ]
}
